shadow/src/Nixfmt/Lexer.hs

```diff
@@ -0,0 +1,97 @@
+{- © 2019 Serokell <hi@serokell.io>
+ - © 2019 Lars Jellema <lars.jellema@gmail.com>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+{-# LANGUAGE LambdaCase, OverloadedStrings #-}
+
+module Nixfmt.Lexer (lexeme) where
+
+import Data.Char
+import Data.List (dropWhileEnd)
+import Data.Maybe (fromMaybe)
+import Data.Text as Text
+  (Text, intercalate, length, lines, null, pack, replace, replicate, strip,
+  stripEnd, stripPrefix, stripStart, takeWhile)
+import Text.Megaparsec hiding (Token, token)
+import Text.Megaparsec.Char
+
+import Nixfmt.Types (Ann(..), Parser, TrailingComment(..), Trivia, Trivium(..))
+import Nixfmt.Util hiding (commonIndentation)
+
+data ParseTrivium
+    = PTNewlines     Int
+    | PTLineComment  Text
+    | PTBlockComment [Text]
+    deriving (Show)
+
+preLexeme :: Parser a -> Parser a
+preLexeme p = p <* manyP (\x -> isSpace x && x /= '\n' && x /= '\r')
+
+newlines :: Parser ParseTrivium
+newlines = PTNewlines <$> Prelude.length <$> some (preLexeme eol)
+
+splitLines :: Text -> [Text]
+splitLines = dropWhile Text.null . dropWhileEnd Text.null
+    . map Text.stripEnd . Text.lines . replace "\r\n" "\n"
+
+stripIndentation :: Int -> Text -> Text
+stripIndentation n t = fromMaybe (stripStart t) $ stripPrefix (Text.replicate n " ") t
+
+commonIndentation :: Int -> [Text] -> Int
+commonIndentation def = foldr min def . map (Text.length . Text.takeWhile (==' '))
```

Comment:
Minor: maybe it's only me, but it seems sad to have two functions with the same name, especially since they are not used only locally in the modules they are defined. Why not call this function e.g. `commonIndentationLength`?